pkg/controller/validatingadmissionpolicystatus/crd.go

```diff
@@ -0,0 +1,253 @@
+		set, ok := t.affectedPolicies[gr]
+		if !ok {
+			set = sets.New[string]()
+			t.affectedPolicies[gr] = set
+		}
+		set.Insert(policyName)
+	}
+}
+
+// observeDependenciesRemoval unregisters the mapping of the policy being deleted.
+func (t *crdTracker) observeDependenciesRemoval(policyName string) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	var dependencies []schema.GroupResource
+	if set, ok := t.policyDependencies[policyName]; ok {
+		dependencies = set.UnsortedList()
+	} else {
+		return
+	}
+	delete(t.policyDependencies, policyName)
+	for _, gvk := range dependencies {
+		if set, ok := t.affectedPolicies[gvk]; ok {
+			set.Delete(policyName)
+			if set.Len() == 0 {
+				delete(t.affectedPolicies, gvk)
+			}
+		}
+	}
+}
+
+func parseCRDVersion(crd *apiextensionsv1.CustomResourceDefinition, crdVersion *apiextensionsv1.CustomResourceDefinitionVersion) schema.GroupVersionKind {
+	gvk := schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: crdVersion.Name,
+		Kind:    crd.Spec.Names.Kind,
+	}
+	return gvk
+}
+
+type cachedTypeResolver struct {
+	meta.RESTMapper
+	resolver.SchemaResolver
+	crdClient apiextensionsv1client.CustomResourceDefinitionInterface
+
+	cachedCRD    map[schema.GroupResource]*apiextensionsv1.CustomResourceDefinition
+	cachedSchema map[schema.GroupVersionKind]*spec.Schema
+}
+
+func newCachedTypeResolver(restMapper meta.RESTMapper, resolver resolver.SchemaResolver, crdClient apiextensionsv1client.CustomResourceDefinitionInterface) *cachedTypeResolver {
+	return &cachedTypeResolver{
+		RESTMapper:     restMapper,
+		SchemaResolver: resolver,
+		crdClient:      crdClient,
+		cachedCRD:      map[schema.GroupResource]*apiextensionsv1.CustomResourceDefinition{},
+		cachedSchema:   map[schema.GroupVersionKind]*spec.Schema{},
+	}
+}
+
+func (r *cachedTypeResolver) ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error) {
+	if s, ok := r.cachedSchema[gvk]; ok {
+		return s, nil
+	}
+	return r.SchemaResolver.ResolveSchema(gvk)
+}
+
+func (r *cachedTypeResolver) KindsFor(resource schema.GroupVersionResource) ([]schema.GroupVersionKind, error) {
+	if isGroupBuiltin(resource.Group) {
+		return r.RESTMapper.KindsFor(resource)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	crd, ok := r.cachedCRD[resource.GroupResource()]
+	if !ok {
+		var err error
+		// FIXME(jiahuif) what if resource is singular?
+		// There does not seem to be a good way to reliably to convert the name without the CRD itself
+		crdName := resource.Resource + "." + resource.Group
+		crd, err = r.crdClient.Get(ctx, crdName, metav1.GetOptions{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				// also keep negative cache
+				r.cachedCRD[resource.GroupResource()] = nil
+				return nil, nil
+			}
+			return nil, err
+		}
+		r.cachedCRD[resource.GroupResource()] = crd
+
+		for _, v := range crd.Spec.Versions {
+			gvk := parseCRDVersion(crd, &v)
+			s, err := convertCRDtoSchema(v.Schema)
```

Comment:
Why are we using CRD schemas instead of using published openapi schemas with the etag?

Reply:
The openapi schema publish is slower to update and we have CRD info handy with the informer added :)

Reply:
- The discovery client takes time to update after CRD change. I considered retry until the new schemas are published, but to figure out if the retrieved schema matches that the informer sees, it would be a convert-and-compare anyway because there is no generation information in the published schemas.
- Anything that involves discovery is expensive especially we need to invalidate caches repeatedly until we see the expected changes.
- Since the controller already holds the subset of the CRDs required to do the type checking and it is already up-to-date, it is better to resolve the paths and schemas locally.